modules/api.py

```diff
@@ -0,0 +1,170 @@
+
+        if self.is_connected():
+            assert self.websocket
+            await self.websocket.close(code=PAPIWebsocketCloseCodes.NORMAL)
+
+        if self.keep_alive_task:
+            try:
+                self.keep_alive_task.cancel()
+            except Exception as e:
+                LOGGER.debug(f'Unable to cancel Pythonista API keep_alive_task in "cog_unload": {e}')
+
+    def is_connected(self) -> bool:
+        return self.websocket is not None and not self.websocket.closed
+
+    async def connect(self) -> None:
+        token: str | None = core.CONFIG["TOKENS"].get("pythonista")
+
+        if not token:
+            self.connection_task = None
+            return
+
+        if self.keep_alive_task:
+            try:
+                self.keep_alive_task.cancel()
+            except Exception as e:
+                LOGGER.debug(f"Failed to cancel Pythonista API Websocket keep alive. This is likely not a problem: {e}")
+
+        headers: dict[str, str] = {"Authorization": token}
+
+        while True:
+            async with aiohttp.ClientSession(headers=headers) as session:
+                try:
+                    self.websocket = await session.ws_connect(url=WS_URL)  # type: ignore
+                except Exception as e:
+                    if isinstance(e, aiohttp.WSServerHandshakeError) and e.status == 403:
+                        LOGGER.critical("Unable to connect to Pythonista API Websocket, due to an incorrect token.")
+                        return
+                    else:
+                        LOGGER.debug(f"Unable to connect to Pythonista API Websocket: {e}.")
+
+                if self.is_connected():
+                    session.detach()
+                    break
+                else:
+                    delay: float = self.backoff.delay()  # type: ignore
+                    LOGGER.debug(f'Retrying Pythonista API Websocket connection in "{delay}" seconds.')
+
+                    await asyncio.sleep(delay)
+
+        self.connection_task = None
+        self.keep_alive_task = asyncio.create_task(self.keep_alive())
+
+    async def keep_alive(self) -> None:
+        assert self.websocket
+
+        initial: dict[str, Any] = {
+            "op": PAPIWebsocketOPCodes.SUBSCRIBE,
+            "subscriptions": [PAPIWebsocketSubscriptions.DPY_MOD_LOG],
+        }
+        await self.websocket.send_json(data=initial)
+
+        while True:
+            message: aiohttp.WSMessage = await self.websocket.receive()
+
+            closing: tuple[aiohttp.WSMsgType, aiohttp.WSMsgType, aiohttp.WSMsgType] = (
+                aiohttp.WSMsgType.CLOSED,
+                aiohttp.WSMsgType.CLOSING,
+                aiohttp.WSMsgType.CLOSE,
+            )
+            if message.type in closing:  # pyright: ignore[reportUnknownMemberType]
+                LOGGER.debug(f"Received a CLOSING/CLOSED/CLOSE message type from Pythonista API.")
```

Comment:
As @chillymosh said, an unnecessary f-string here!
```suggestion
                LOGGER.debug("Received a CLOSING/CLOSED/CLOSE message type from Pythonista API.")
```

Reply:
Umbra and Chilly, good catch,